### ISSUE TITLE
In a short trade, `stop_loss_abs` is lower than `initial_stop_loss_abs`, meanwhile `stop_loss_ratio` is lower than `initial_stop_loss_ratio`

### ISSUE THREAD

**Issue by funk1d** [freqtrade/freqtrade#9125]
## Describe your environment

  * Operating system: Debian 11
  * Python Version: 3.11.3 (`python -V`)
  * CCXT version: 4.0.36 (`pip freeze | grep ccxt`)
  * Freqtrade Version: 2023.7 (`freqtrade -V` or `docker compose run --rm freqtrade -V` for Freqtrade running in docker)
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

```json
{
    "pair": "BTC/USDT:USDT",
    "stake_amount": 19.80816,
    "max_stake_amount": 1980.816,
    "amount": 0.096,
    "open_date": "2022-06-15 17:30:00+00:00",
    "close_date": "2022-06-15 17:34:00+00:00",
    "open_rate": 20633.5,
    "close_rate": 20705.0,
    "fee_open": 0.0002,
    "fee_close": 0.0002,
    "trade_duration": 4,
    "profit_ratio": -0.386670495705267,
    "profit_abs": -7.6576992,
    "exit_reason": "trailing_stop_loss",
    "initial_stop_loss_abs": 20829.5,
    "initial_stop_loss_ratio": -0.95,
    "stop_loss_abs": 20705.0,
    "stop_loss_ratio": -1.0,
    "min_rate": 20472.2,
    "max_rate": 20742.5,
    "is_open": false,
    "enter_tag": null,
    "leverage": 100.0,
    "is_short": true,
    "open_timestamp": 1655314200000,
    "close_timestamp": 1655314440000,
    "orders": [
      {
        "amount": 0.096,
        "safe_price": 20633.5,
        "ft_order_side": "sell",
        "order_filled_timestamp": 1655314200000,
        "ft_is_entry": true
      },
      {
        "amount": 0.096,
        "safe_price": 20705.0,
        "ft_order_side": "buy",
        "order_filled_timestamp": 1655314440000,
        "ft_is_entry": false
      }
    ]
  }
```
i'm doing backtest, and this short trade exits because of `trailing_stop_loss`
its `stop_loss_ratio` is `-1.0`, but its `stop_loss_abs` is lower than `initial_stop_loss_abs`
i'm wondering how does the `stop_loss_abs` get calculated?

**Comment by froggleston** [comment]
What's your trailing stop config, or custom_stoploss callback?

**Comment by funk1d** [comment]
i disable the `trailing_stop` and use `custom_stoploss`
```python
trailing_stop = False

def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
    dataframe["atr"] = ta.ATR(dataframe, timeperiod=14)

def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                    current_rate: float, current_profit: float, **kwargs) -> float:
    dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
    trade_date = timeframe_to_prev_date(self.timeframe, trade.open_date_utc)
    trade_candle = dataframe.loc[dataframe['date'] == trade_date]
    if not trade_candle.empty:
        trade_candle = trade_candle.squeeze()
        sign = -1 if not trade.is_short else 1
        return stoploss_from_absolute(trade.open_rate + sign * 3.0 * trade_candle['atr'],
                                        current_rate,
                                        is_short=trade.is_short) * trade.leverage
    # return maximum stoploss value, keeping current stoploss price unchanged
    return 1
```

**Comment by perkmeister** [comment]
Not sure this bit makes sense:
```
return stoploss_from_absolute(trade.open_rate + sign * 3.0 * trade_candle['atr'],
                                        current_rate,
                                        is_short=trade.is_short) * trade.leverage
```
Edit: I'm not sure you should be adusting for leverage here, because returning for example 0.1 with 10x leverage it would be turned into 0.01 when the stoploss gets set. This to me makes more sense:  
```
new_stoploss =  trade.open_rate + (sign * 3.0 * trade_candle['atr'])
return stoploss_from_absolute(new_stoploss, current_rate, is_short=trade.is_short)
```

**Comment by xmatthias** [comment]
Based on the leverage of 100x (insane) - one problematic point is the fallback to `return 1` (maybe not the only one, didn't look in too much detail at the rest).

This will result in a trailing stoploss of 1 / 100 (1%) once your trade reaches "some" profit - turning the stop into a trailing stop at 100% capital (which at 100x, is 1% price movements).

It's unlikely you'll have hit that code path in live (unless you restart the bot i think) - but (especially in backtesting) still a possibility.

**Comment by funk1d** [comment]
> Not sure this bit makes sense:
> 
> ```
> return stoploss_from_absolute(trade.open_rate + sign * 3.0 * trade_candle['atr'],
>                                         current_rate,
>                                         is_short=trade.is_short) * trade.leverage
> ```
> 
> Edit: I'm not sure you should be adusting for leverage here, because returning for example 0.1 with 10x leverage it would be turned into 0.01 when the stoploss gets set. This to me makes more sense:
> 
> ```
> new_stoploss =  trade.open_rate + (sign * 3.0 * trade_candle['atr'])
> return stoploss_from_absolute(new_stoploss, current_rate, is_short=trade.is_short)
> ```

Thanks for your advice, but I think that's the why I have to multiply `trade.leverage` at the end in order to avoid the "returning for example 0.1 with 10x leverage it would be turned into 0.01 when the stoploss gets set" scenario

**Comment by funk1d** [comment]
> Based on the leverage of 100x (insane) - one problematic point is the fallback to `return 1` (maybe not the only one, didn't look in too much detail at the rest).
> 
> This will result in a trailing stoploss of 1 / 100 (1%) once your trade reaches "some" profit - turning the stop into a trailing stop at 100% capital (which at 100x, is 1% price movements).
> 
> It's unlikely you'll have hit that code path in live (unless you restart the bot i think) - but (especially in backtesting) still a possibility.

Yes, I guess that fallback line is the reason why the `stop_loss_ratio` turns to `-1.0`, but I'm still confuse about: how does the `"stop_loss_abs": 20705.0` get calculated?

**Comment by perkmeister** [comment]
> Thanks for your advice, but I think that's the why I have to multiply trade.leverage at the end in order to avoid the "returning for example 0.1 with 10x leverage it would be turned into 0.01 when the stoploss gets set" scenario

I still don't think this is right. I mean if you say wanted 20% TP and 10% SL the `current_profit` value would also be multiplied by leverage, so in reality a 20% profit would be a 2% increase in price, but you'd be setting the real stoploss to 10% if you multiplied it and not 1%. Also if you were to do that then you shouldn't be multiplying the result returned by `stoploss_from_absolute()` as that is a ratio relative to current_rate, I think it should be scaled as an offset to `open_rate`, like this (but again I don't think you should be doing it). Happy to be proved wrong though:

```
new_stoploss =  trade.open_rate + (sign * 3.0 * trade_candle['atr'] * trade.leverage)
return stoploss_from_absolute(new_stoploss, current_rate, is_short=trade.is_short)
```

**Comment by funk1d** [comment]
> I still don't think this is right. I mean if you say wanted 20% TP and 10% SL the `current_profit` value would also be multiplied by leverage, so in reality a 20% profit would be a 2% increase in price, but you'd be setting the real stoploss to 10% if you multiplied it and not 1%. Also if you were to do that then you shouldn't be multiplying the result returned by `stoploss_from_absolute()` as that is a ratio relative to current_rate, I think it should be scaled as an offset to `open_rate`, like this (but again I don't think you should be doing it). Happy to be proved wrong though:
> 
> ```
> new_stoploss =  trade.open_rate + (sign * 3.0 * trade_candle['atr'] * trade.leverage)
> return stoploss_from_absolute(new_stoploss, current_rate, is_short=trade.is_short)
> ```

Actually what I really want to do is: set an absolute stop loss price, and make sure this stop loss price would not be affected by the `trade.leverage`, so I multiply the `trade.leverage` at the end.
I'm not sure if this is correct?

**Comment by perkmeister** [comment]
I'm really not sure, my feeling is because `stoploss_from_open()` would be using the `current_profit`, which in turn relies on `current_rate` and `open_rate`, that the scaling should be as an offset to `open_rate` but I could be wrong. I suggest asking the question in Discord.

**Comment by xmatthias** [comment]
I find the "leverage" adjustment applied correctly.
The return value is a the "acceptable risk" for the position - so if the return value is 5% - it'll result in an effective 1% price change - so if you want "stop from absolute" - then you need to multiply by leverage - to counter this same move on freqtrade internals.


it's also pretty easy to simulate and visualize in dry-run...

```
        dataframe['atr'] = ta.ATR(dataframe, timeperiod=14)
        offset = 3
        dataframe['atr_up'] = dataframe['open'] + dataframe['atr'] * offset
        dataframe['atr_down'] = dataframe['open'] - dataframe['atr'] * offset
```

You can then configure your plot to actually show these lines.

I then simply opened a 5x short position.

![image](https://github.com/freqtrade/freqtrade/assets/5024695/c4fc55fd-6971-4d60-a51b-03e8449af386)

Not multiplying by leverage results in the stop being too tight:

![image](https://github.com/freqtrade/freqtrade/assets/5024695/6063132d-71c9-4548-869d-8d1e4b470db3)


(to be fair, both images have been taken with an atr multiplier of 10 - as otherwise the stop triggered too fast on a 5x position.

---

You'll also encounter the "fall-through" to 1 immediately, as the "trade open candle" does not exist for the duration of the opening candle (that candle isn't closed, therefore not available yet).
You'll therefore need to return -100 or None (assuming you use the lastest develop) to avoid having the "-1" adjust your plotting.

**Comment by moinonin** [comment]
I have been having the same issue. Assuming one gets a good strategy but not leveraged and you want to run it as leveraged. Could someone present for example a function that takes leverage value and returns adjusted stoploss? and any other value that must change when leverage changes from say 1.0x to 5.0x. Highly appreciated.

**Comment by xmatthias** [comment]
@moinonin what your asking for is way too broad.

We won't know HOW your actual stop works, or what exact calculations and methods it implies.

if you're about the "risk involved" (say, i wanna risk 10% of my investment) - then nothing will actually change - as the leverage adjustment will account for that - moving your stop to trigger after a 2% price movement).
For absolute values - (like in this case) - the adjustment will have to be made.

For that reason, the helper method will now include a `leverage` parameter, which will take care of this calculation.
So effectively, usage can be

``` python
        return stoploss_from_absolute(trade.open_rate + sign * 3.0 * trade_candle['atr'],
                                        current_rate,
                                        is_short=trade.is_short, leverage=trade.leverage)
```
